### ISSUE TITLE
[Bug]: `FetchPlugin` does not support server-side rendering

### ISSUE THREAD

**Issue by CharlesStover** [aws-observability/aws-rum-web#415]
### Which web client version did you detect this bug with?

v1.12.0

### What environment (build systems, module system, and framework) did you detect this bug with?

- NextJS 12.3.4
- React 18.2.0
- TypeScript 4.9.5
- Webpack (unknown version, NextJS internal)

### Is your web application a single page application (SPA) or multi page application (MPA)?

SPA

### Please provide your web client configuration

N/A

### Please describe the bug/issue

When instantiating `new AwsRum()`, server-side rendering throws an error:

```
TypeError: Failed to fetch
     at VirtualPageLoadTimer._this.fetch (webpack-internal:///./node_modules/aws-rum-web/dist/es/sessions/VirtualPageLoadTimer.js:57:18)
     at eval (webpack-internal:///./node_modules/aws-rum-web/dist/es/sessions/VirtualPageLoadTimer.js:71:33)
     at FetchPlugin._this.fetch (webpack-internal:///./node_modules/aws-rum-web/dist/es/plugins/event-plugins/FetchPlugin.js:163:18)
     at eval (webpack-internal:///./node_modules/aws-rum-web/dist/es/plugins/event-plugins/FetchPlugin.js:179:33)
```

It appears to be trying to call `window.fetch` once instantiated, which of course does not exist.

I would like to instantiate `new AwsRum` on the server without it actually connecting to emit any metrics, since there is no user in this context -- there should be nothing to emit, as no events have been triggered.

* I understand that a server-side rendered page cannot (should not) be connecting to RUM to emit metrics.
* My concern is that I do not believe that instantiating `AwsRum` _in and of itself_ should not be connecting to RUM to emit metrics.

Please advise. 😊

**Comment by adebayor123** [comment]
Hi Charles, web clients should not be initialized on a server - is there a specific reason why you wish to do so?

**Comment by CharlesStover** [comment]
I'm server rendering a React application, so the client side runtime code is also executing on the server.

**Comment by hjniermann** [comment]
Hi, does this issue mean that RUM is not supported when server-side rendering is enabled on frameworks such as NextJS? This should be clearly noted, as NextJS is currently the recommended React framework, and is a blocker to teams using server-side rendering. @CharlesStover , did you find a way around this for your use case?

**Comment by CharlesStover** [comment]
You can use `useEffect` to instantiate RUM only after your application has loaded on the client, even though this is an anti-pattern per NextJS.

You can use [`patch-package`](https://www.npmjs.com/package/patch-package) to patch the RUM package to not call `window.fetch` on the server.

I forget which path I ended up going.

**Comment by qhanam** [comment]
When initialized, the web client (1) performs steps to get credentials, and (2) emits a session start event. I believe both of these events can be deferred.

Regarding (2), initialize the web client in the disabled state by setting `enabled: false` in the config. The web client will not dispatch events when disabled.

This setting is currently missing from [the docs](https://github.com/aws-observability/aws-rum-web/blob/main/docs/configuration.md), which we should fix. The use of this setting is demonstrated in [this blog](https://aws.amazon.com/blogs/mt/how-to-isolate-signed-in-users-from-guest-users-within-amazon-cloudwatch-rum/).

Regarding (1), omit the `guestRoleArn` and `identityPoolId` from the config. When the application is ready, use the `setAwsCredentials` command to pass the credential provider.

We may want to do something so that credentials are not fetched when the web client is disabled, in which case the application need only flip the web client from disabled to enabled when it is ready.

**Comment by qhanam** [comment]
Please re-open if this doesn't work or if you have thoughts on a better solution.

Meanwhile, we'll add these to the backlog.

**Comment by CharlesStover** [comment]
I'll give those a try. As a placeholder, I've seen as a common pattern for SSR+CSR packages to use `if (typeof window === 'undefined')` to determine if they are rendering on a client or server. This pattern may work for you if you want RUM to no-op on a server render: only call window.fetch if window is defined.

**Comment by qhanam** [comment]
So to clarify, when rendering on a server (i.e., no window object), you expect all calls to the web client to be no-ops?

**Comment by CharlesStover** [comment]
>So to clarify, when rendering on a server (i.e., no window object), you expect all calls to the web client to be no-ops?

Yes, server-rendering serves primarily to generate the initial DOM. Once passed to the client, React will "hydrate," re-executing the render to (1) build the virtual DOM into memory [since the actual DOM is already present] and (2) execute any non-DOM side effects. For example, instantiating RUM and monitoring web vitals does not impact the initial DOM or view, so this behavior does not need to occur during the initial server render. There's no value in having a RUM object in server memory, but once the application re-hydrates on the client, RUM can instantiate in memory and perform browser-based actions like fetching and web vitals.

For a server-side rendered application, the application code is executed once on the server to build the initial DOM and again on the client to hydrate the application into browser memory. By saying "only perform this actions when the `window` object exists," it will no-op on the server, but _it will still execute_ in the browser during the hydration cycle of the application, as that same JavaScript will re-execute, however now in an environment where `window` exists.

Thanks for the attention to the issue. 🙂

**Comment by moltar** [comment]
Agree with @CharlesStover here. Would be great to have a mock/stub implementation of the RUM library.

Not only for SSR purposes but also for testability.

A great example of this mocking ability is Undici:

https://undici.nodejs.org/#/docs/api/MockClient

This greatly simplifies testing, and don't need to monkey patch or use `jest.mock`.

**Comment by CharlesStover** [comment]
@moltar If you are using React, my package [`aws-rum-react`](https://www.npmjs.com/package/aws-rum-react) comes with a `MockAwsRumProvider` that you can wrap your test in to mock implementation of the RUM library. It still doesn't support SSR per se, but it can help with unit tests. :)

**Comment by sonu-gameskraft** [comment]
Core functionality of RUM for SSR website is not working to capture http error.

1. Upgraded the RUM version from 1.18.0 to 1.19.0
2. I would like to clarify that the Performance API focuses on performance-related metrics, which don't meet my request. I only want to track HTTP and error metrics. But still, I tried implementing the script you suggested for the Performance API, but unable to catch the 403 error.
3. I've increased the priority of the cwr.js file, so now it's loaded before the generated JS file. However, we are still experiencing issues with session and not catching HTTP errors.
4. performance telemetry also not able to capture 4xx and 5XX error 

<img width="1728" alt="Screenshot 2024-08-22 at 2 38 52 PM" src="https://github.com/user-attachments/assets/8792d5fa-1871-49c8-9253-7ccdcde58180">
<img width="1728" alt="Screenshot 2024-08-22 at 2 38 52 PM" src="https://github.com/user-attachments/assets/5a7a15a4-2da1-4d9c-a328-c3a0537ae96f">
